### ISSUE TITLE
Esp32 causing data corruption when delay is introduced in the first time?

### ISSUE THREAD

**Issue by shariq-azim** [espressif/arduino-esp32#8291]
### Board

esp32 dev module

### Device Description

Its not connected to anything

### Hardware Configuration

Currently no hardware is connected

### Version

v2.0.9

### IDE Name

arduino ide 2.0

### Operating System

windows 11

### Flash frequency

80Mhz

### PSRAM enabled

no

### Upload speed

921600

### Description

Hello Everyone,
I am facing this weird issue where due to some reasons. The **delay** is causing issues with the free heap memory which indirectly is causing data issues. The following is the part in void loop of the sketch. **Releasing something around 4kb** and causing data corruption to a variable of **type const char*** which is the type accepted in pubsubclient's mqtt server parameter

I was expecting it to print IP which is printed before the delay is introduced.

The complete code is much much bigger and have not been excluded fully

### Sketch

```cpp
#include <WiFi.h>
#include <WiFiClient.h>
//#include <WebServer.h>
#include <ESPmDNS.h>
#include <Update.h>

#include <PubSubClient.h>
//#include <HTTPClient.h>
#include <ArduinoJson.h>
#include <string.h>
#include <cstring>

#include <WiFiManager.h>
#include <Preferences.h>

#define DEBUG_SW 1

const char* dnshost = "shariqESP";
char ssid[25] = "xyz xyz";

char pass[20] = "alpha beta";

String ssidPref;
String passPref;

char* mqttServer = "192.168.1.60";  //check actual config of node-red
//const int mqttPort = 1886;
char mqttPort[6] = "8886";
const char* mqttUser = "";
const char* mqttPassword = "";

bool InternetStatus = false;

uint32_t chipId = 0;    //for chip id
String chipIdStr = "";  //for chip id
String topicOTA = "ESP32/OTAStatus/";
String topicStatus = "ESP32/Status/";
String PiStatus = "PiServer/";
String initTopic = "PiServer/ESP32/start";
String PiOTAInpStatus = "PiServer/ESP32/OTA/";

String ChipId = "";

long lastReconnectAttempt = 0;

String mqttServerPref;
String mqttPortPref;
WiFiClient espClient;

PubSubClient pubsubclient(espClient);
Preferences preferences;


void without_internet() {
  //some logic
}

void with_internet() {
  //some more logic
}

void saveConfigCallback() {
  Serial.println("saveConfigCallback  fired");
}

void WiFiStationConnected(WiFiEvent_t event, WiFiEventInfo_t info) {
  Serial.println("Connected to AP successfully!");
}

void WiFiGotIP(WiFiEvent_t event, WiFiEventInfo_t info) {
  Serial.println("WiFi connected");
  Serial.println("IP address: ");
  Serial.println(WiFi.localIP());
}

void WiFiStationDisconnected(WiFiEvent_t event, WiFiEventInfo_t info) {
  Serial.println("Disconnected from WiFi access point");
  Serial.print("WiFi lost connection. Reason: ");
  //Serial.println(info.disconnected.reason);
  Serial.println(info.wifi_sta_disconnected.reason);
  Serial.println("Trying to Reconnect");
  WiFi.begin(ssid, pass);
  //WiFi.reconnect();
}
void callback(char* topic, byte* payload, unsigned int length) {
  //some logic
}

void setup() {

  Serial.begin(115200);
  without_internet();

  delay(10);
  for (int i = 0; i < 17; i = i + 8) {  //get chip id for esp32
    chipId |= ((ESP.getEfuseMac() >> (40 - i)) & 0xff) << i;
  }

  chipIdStr = String(chipId).substring(0, 5);
  ChipId = chipIdStr;
  topicOTA += chipIdStr;
  topicStatus += chipIdStr;


  Serial.print("OTA: ");
  Serial.println(topicOTA.c_str());
  Serial.print("Status: ");
  Serial.println(topicStatus.c_str());
  preferences.begin("credentials", false);

  //WiFiManager
  //Local intialization. Once its business is done, there is no need to keep it around

  WiFiManagerParameter custom_mqtt_server("server", "mqtt server", mqttServer, 40);
  WiFiManagerParameter custom_mqtt_port("port", "mqtt port", mqttPort, 5);
  WiFiManagerParameter custom_field;  // global param ( for non blocking w params )

  WiFiManager wm;

  wm.setConfigPortalTimeout(40);
  //set config save notify callback
  wm.setSaveConfigCallback(saveConfigCallback);

  //wm.setDebugOutput(true);
  //add all your parameters here
  wm.addParameter(&custom_mqtt_server);
  wm.addParameter(&custom_mqtt_port);

  String html_script = "<script>document.getElementById('s').required = true;document.getElementsByName('s')[0].labels[0].innerHTML=document.getElementsByName('s')[0].labels[0].innerHTML +\"<span style='color: red;'>*</span>\";";
  html_script += "document.getElementById('p').required = true;document.getElementsByName('p')[0].labels[0].innerHTML=document.getElementsByName('p')[0].labels[0].innerHTML +\"<span style='color: red;'>*</span>\";";
  html_script += "document.getElementById('server').required = true;document.getElementsByName('server')[0].labels[0].innerHTML=document.getElementsByName('server')[0].labels[0].innerHTML +\"<span style='color: red;'>*</span>\";";
  html_script += "document.getElementById('port').required = true;document.getElementsByName('port')[0].labels[0].innerHTML=document.getElementsByName('port')[0].labels[0].innerHTML +\"<span style='color: red;'>*</span>\";";
  html_script += "</script>";

  new (&custom_field) WiFiManagerParameter(html_script.c_str());  // custom html input
  wm.addParameter(&custom_field);

  if (!wm.autoConnect("AutoConnectAP")) {
    Serial.println("failed to connect and hit timeout");
    delay(3000);
  }


  String saved_ssid_old = wm.getWiFiSSID().c_str();
  String saved_pass_old = wm.getWiFiPass().c_str();

  String saved_ssid = WiFi.SSID().c_str();
  Serial.print("SSID1: ");
  Serial.println(saved_ssid);


  String saved_pass = WiFi.psk().c_str();

  Serial.println(custom_mqtt_server.getValue());
  Serial.println(custom_mqtt_port.getValue());

  preferences.begin("MQTTAddress", false);

  //if( custom_mqtt_server.getValue() && !custom_mqtt_server.getValue()[0]){
  if (strlen(custom_mqtt_server.getValue()) == 0) {
    Serial.println("NO MQTT server not provided,reading from preference");
    mqttServerPref = preferences.getString("mqtt_server", "");
    if (mqttServerPref == "") {
      Serial.println("No values saved for mqttServer in Prefences");
    }
    Serial.println(mqttServerPref);
    strcpy(mqttServer, mqttServerPref.c_str());

  }

  else if (strlen(custom_mqtt_server.getValue()) != 0) {
    //strcpy(mqttServer , (char*)custom_mqtt_server.getValue());
    mqttServer = (char*)custom_mqtt_server.getValue();  //because datatype of mqttServer is char*
    String mqttServerPref = preferences.getString("mqtt_server", "");
    if (strcmp(mqttServer, mqttServerPref.c_str()) != 0) {
      preferences.putString("mqtt_server", custom_mqtt_server.getValue());
      Serial.println("MQTT settings[Server] saved using Preferences");
    }
  }


  //PORT
  if (strlen(custom_mqtt_port.getValue()) == 0) {
    Serial.println("NO MQTT port not provided,reading from preference");

    mqttPortPref = preferences.getString("mqtt_port", "");
    if (mqttPortPref == "") {
      Serial.println("No values saved for mqttPort in Prefences");
    }
    Serial.println(mqttPortPref);
    strcpy(mqttPort, mqttPortPref.c_str());
  }

  else if (strlen(custom_mqtt_port.getValue()) != 0) {
    strcpy(mqttPort, custom_mqtt_port.getValue());
    String mqttPortPref = preferences.getString("mqtt_port", "");

    if (strcmp(mqttPort, mqttPortPref.c_str()) != 0) {
      preferences.putString("mqtt_port", custom_mqtt_port.getValue());
      Serial.println("MQTT settings[Port] saved using Preferences");
    }
  }

  preferences.end();


  Serial.print("MQTT IP: ");
  Serial.println(mqttServer);
  Serial.print("MQTT IP: ");
  Serial.println(mqttPort);

  preferences.begin("credentials", false);

  if ((strlen(WiFi.SSID().c_str()) == 0) || (WiFi.SSID().c_str()[0] == '\0') || (WiFi.SSID().c_str() == "null")) {

    ssidPref = preferences.getString("P_ssid", "");
    if (ssidPref == "") {
      Serial.println("No values saved for ssid in Prefences");
    }

    //Serial.println( ssidPref);

    strcpy(ssid, ssidPref.c_str());


  }

  else if (WiFi.SSID().c_str() != "") {
    // Length (with one extra character for the null terminator)
    int str_len = saved_ssid.length() + 1;
    // Prepare the character array (the buffer)
    char char_array[str_len];

    // Copy it over
    saved_ssid.toCharArray(char_array, str_len);
    strcpy(ssid, char_array);

    preferences.putString("P_ssid", ssid);

    ssidPref = preferences.getString("P_ssid", "");
    if (strcmp(ssid, ssidPref.c_str()) != 0) {
      preferences.putString("P_ssid", ssid);
    }
  }

  //password
  if ((strlen(WiFi.psk().c_str()) == 0) || (WiFi.psk().c_str()[0] == '\0') || (WiFi.psk().c_str() == "null")) {
    passPref = preferences.getString("P_pass", "");
    if (passPref == "") {
      Serial.println("No values saved for password in Prefences");
    }
    strcpy(pass, passPref.c_str());

  }

  else if (WiFi.psk().c_str() != "") {
    int str_len2 = saved_pass.length() + 1;
    // Prepare the character array (the buffer)
    char char_array2[str_len2];

    // Copy it over
    saved_pass.toCharArray(char_array2, str_len2);
    strcpy(pass, char_array2);


    passPref = preferences.getString("P_pass", "");
    if (strcmp(pass, passPref.c_str()) != 0) {
      preferences.putString("P_pass", pass);
    }
  }
  preferences.end();

  wm.disconnect();
  delay(1000);

  WiFi.onEvent(WiFiStationConnected, WiFiEvent_t::ARDUINO_EVENT_WIFI_STA_CONNECTED);
  WiFi.onEvent(WiFiGotIP, WiFiEvent_t::ARDUINO_EVENT_WIFI_STA_GOT_IP);
  WiFi.onEvent(WiFiStationDisconnected, WiFiEvent_t::ARDUINO_EVENT_WIFI_STA_DISCONNECTED);

  WiFi.begin(ssid, pass);
  int ct = 0;

  while (WiFi.status() != WL_CONNECTED) {
    delay(1000);
    ct = ct + 1000;
    if (ct > 5000) {
      Serial.println("not able to connect to wifi");
      break;
    }
  }
  Serial.print("Heap size before mqtt check: ");
  Serial.println(ESP.getFreeHeap());

  InternetStatus = setup_mqtt();  //0.5

  Serial.print("Heap size exit setup: ");
  Serial.println(ESP.getFreeHeap());
}

bool setup_mqtt() {
  preferences.begin("MQTTAddress", false);
  String mqttServerPref = preferences.getString("mqtt_server", "");


  preferences.end();
  long now = millis();
  //long lastReconnectAttempt = 0;

  bool connectedBool = false;
  Serial.print("Setting mqtt details: ");
  Serial.print(mqttServer);
  Serial.print(" ");
  Serial.print(mqttServerPref);
  Serial.print(" ");
  Serial.println(atoi(mqttPort));

  pubsubclient.setServer(mqttServer, atoi(mqttPort));  //v.5d
  pubsubclient.setCallback(callback);
  int mqct = 0;
  if (pubsubclient.connected()) {

    return true;
  } else {
    while ((now - lastReconnectAttempt) > 5000) {
      lastReconnectAttempt = millis();
      Serial.println("Connecting to MQTT...");
      if (pubsubclient.connect(chipIdStr.c_str())) {
        //if (client.connect("ESP32Client")) {

        lastReconnectAttempt = 0;

        Serial.println("connected");


        pubsubclient.subscribe(PiStatus.c_str());        //"PiServer/xyz/Status"
        pubsubclient.subscribe(PiOTAInpStatus.c_str());  //"PiServer/ESP/OTA/xyz"
        Serial.print("Trying to publish in ");
        Serial.println(topicStatus.c_str());
        pubsubclient.publish(initTopic.c_str(), ChipId.c_str());  //send chip id when connected
        delay(1000);
        pubsubclient.publish(topicStatus.c_str(), "ESP device connected ");

        StaticJsonDocument<200> doc;

        JsonObject obj = doc.createNestedObject("output");
        JsonObject light = obj.createNestedObject("light");

        obj["fan"] = "2";

        char topicbuffer[256];
        size_t n = serializeJson(doc, topicbuffer);
        pubsubclient.publish(topicStatus.c_str(), topicbuffer, n);
        Serial.println(topicbuffer);
        doc.clear();

        connectedBool = true;
        return connectedBool;
      }
      Serial.print("failed with state ");
      Serial.println(pubsubclient.state());

      if (connectedBool) {
        connectedBool = false;
      }
    }
  }

  if (!connectedBool) {
    Serial.println("MQTT cannot be reached");
  }

  Serial.print("Exiting mqtt details: ");
  Serial.println(mqttServer);
  return connectedBool;
}



void loop() {
  Serial.print("Heap size start loop: ");
  Serial.println(ESP.getFreeHeap());
  Serial.println(mqttServer);
  Serial.println(mqttPort);


  fanPinChange = (bool)((fanPin1 != digitalRead(s1)) || (fanPin2 != digitalRead(s2)) || (fanPin3 != digitalRead(s3)) || (fanPin4 != digitalRead(s4)));

  if (fanPin1 != digitalRead(s1)) {
    fanPin1 = digitalRead(s1);
  }
  if (fanPin2 != digitalRead(s2)) {
    fanPin2 = digitalRead(s2);
  }
  if (fanPin3 != digitalRead(s3)) {
    fanPin3 = digitalRead(s3);
  }
  if (fanPin4 != digitalRead(s4)) {
    fanPin4 = digitalRead(s4);
  }  

  Serial.print("Heap size check loop1: ");
  Serial.println(ESP.getFreeHeap());
  Serial.println(mqttServer);

  Serial.println(WiFi.status());
  Serial.print("Heap size check loop2: ");
  Serial.println(ESP.getFreeHeap());
  Serial.println(mqttServer);
  //no issues so far



  delay(1000);  //**notice from here**
  Serial.print("Heap size check loop3: ");
  Serial.println(ESP.getFreeHeap());
  Serial.println(mqttServer);
  //prints gibberish/corrupted data

  //Check WiFi connection status
  Serial.print("Wifi status: ");


  if (WiFi.status() != WL_CONNECTED) {  // you can skip to the below else part for now
    //digitalWrite(R6, LOW);
    if (DEBUG_SW) Serial.println("Wifi Not Connected");
    delay(400);  //custom
    delay(400);  //custom
    Serial.print("Heap size check loop wifi off: ");
    Serial.println(ESP.getFreeHeap());
    Serial.println(mqttServer);

  } else {
    Serial.println("Wifi Connected");
    Serial.print("Heap size after wifi connected and before delay: ");
    Serial.println(ESP.getFreeHeap());
    Serial.println(mqttServer);
    delay(1000);
    //if (DEBUG_SW)
    Serial.println("Wifi Connected");
    Serial.print("Heap size after wifi connected: ");
    Serial.println(ESP.getFreeHeap());
    Serial.println(mqttServer);  //prints gibberish/corrupted data
    //0.4 logic added
    //reconnect to mqtt or process if incoming string present
    if (pubsubclient.connected()) {
      Serial.println("connected with MQTT");
     // digitalWrite(R6, HIGH);
      Serial.print("Heap size check wifi on pubsub connected: ");
      Serial.println(ESP.getFreeHeap());
      Serial.println(mqttServer);

      pubsubclient.loop();
      Serial.print("Heap size check LOOP wifi on pubsub connected: ");
      Serial.println(ESP.getFreeHeap());
      Serial.println(mqttServer);
      InternetStatus = true;
      Serial.print("Heap size check LOOP wifi on pubsub connected2: ");
      Serial.println(ESP.getFreeHeap());
      Serial.println(mqttServer);

    } else {
      Serial.print("Heap size after pubsub: ");
      Serial.println(ESP.getFreeHeap());
      Serial.println(mqttServer);
      Serial.print("Reconnect to MQTT server failed with state ");
      Serial.println(pubsubclient.state());
      //digitalWrite(R6, LOW);
      InternetStatus = setup_mqtt();
    }
  }

  Serial.print("Heap size before internetstatus: ");
  Serial.println(ESP.getFreeHeap());

  //not important here

  if (InternetStatus)
    with_internet();
  else
    without_internet();
  Serial.print("Heap size on exit loop: ");
  Serial.println(ESP.getFreeHeap());
}
```

### Debug Message

```plain
printed from 
//  Serial.print("Heap size check loop1: ");
//  Serial.println(ESP.getFreeHeap());
//  Serial.println(mqttServer);

192.168.1.60
3
Heap size check loop2: 252872
192.168.1.60
Heap size check loop3: 256044
 ��?6�
Wifi status: Wifi Connected
Heap size after wifi connected and before delay: 256044
 ��?6�
Wifi Connected
Heap size after wifi connected: 256044
 ��?�
connected with MQTT
Heap size check wifi on pubsub connected: 256044
 ��?�
Heap size check LOOP wifi on pubsub connected: 256604

```

![image](https://github.com/espressif/arduino-esp32/assets/37748914/22547e0b-bbd9-4740-aa49-fb0dcdc470ff)

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

**Comment by shariq-azim** [comment]
The code could be overwhelming, but the issue is observed in the logs of void loop...

**Comment by everslick** [comment]
You are walking on thin ice by doing something like `strcpy(mqttServer, mqttServerPref.c_str());` while mqttServer is pointing to a `const char *`. You are generally messing with strcpy while dealing mostly with Strings, so I bet there is some mem corruption or a dangling pointer lurking in your code.

**Comment by shariq-azim** [comment]
Whats your suggestion then @everslick ?

**Comment by shariq-azim** [comment]
> ```c++
> if
> ```

Part  of  Void loop..
Just after the delay is introduced  before the
WiFi.status() != WL_CONNECTED part @ibernstone

**Comment by aeonSolutions** [comment]
I had to spent some additional time debugging my smart hardware electronics for a home automation wall mount switch based on the ESP32 S3 ( see it on my GitHub [here](https://github.com/aeonSolutions/AeonLabs-Smart-Power-Switch-Outlet/blob/main/README.md) ) 

[![enter image description here][1]][1]

because I was able to upload the firmware code into it however when reading serial output it only displayed "gibberish" characters. 

On the internet I could not find the answer, so I had to work harder to find the solution. Common problems found when dealing with serial output errors are:
- poor UART ground
- absence of a 500 Ohm resistor on the TX line
- UART cable to long
- track coupling and track length mismatch

are all hardware issues that cause serial communication errors and malfunctioning. However, for the ESP32 there's one missing on the list above. 
On this particular hardware, I had incorrectly soldered the RTC crystal of 32,768MHz into the MCU crystal place which is for one of 40MHz. The result was, it is possible to do firmware upload using a 32,768Mhz crystal for the MCU however, and due to the incorrect clock frequency, serial output is no longer 115200bps and displays "gibberish" characters.   

**In summary make sure the crystal has the correct clock frequency value**.




  [1]: https://i.stack.imgur.com/wACFh.png
